platform/darwin/src/darwin_conversion.hpp

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <mbgl/util/logging.hpp>
+#include <mbgl/style/conversion.hpp>
+#include <mbgl/util/feature.hpp>
+#include <mbgl/util/optional.hpp>
+
+namespace mbgl {
+namespace style {
+namespace conversion {
+
+// A minimal wrapper class conforming to the requirements for `objectMember(v, name)` (see mbgl/style/conversion.hpp)
+// This is necessary because using `NSObject*` as the value type in `optional<NSObject*>` causes problems for the ARC,
+// due to things like `optional(const value_type& __v)`
+class OptionalNSObjectValue {
+public:
+    OptionalNSObjectValue(NSObject *_value) : value(_value) {}
+    OptionalNSObjectValue() {}
+    
+    explicit operator bool() const {
+        return value;
+    }
+    
+    NSObject* operator*() {
+        assert(this);
+        return value;
+    }
+private:
+    NSObject* value;
+};
+
+inline bool isUndefined(const NSObject* value) {
+    return !value || value == [NSNull null];
+}
+
+inline bool isArray(const NSObject* value) {
+    return [value isKindOfClass:[NSArray class]];
+}
+
+inline bool isObject(const NSObject* value) {
+    return [value isKindOfClass:[NSDictionary class]];
+}
+
+inline std::size_t arrayLength(const NSObject* value) {
+    NSCAssert([value isKindOfClass:[NSArray class]], @"Value must be an NSArray for getLength().");
+    NSArray * array = (NSArray *)value;
```

Comment:
I’d still encourage you to explore the use of `id` instead of `NSObject *`. Saying “NSObject” here tells the compiler (and the code reviewer) that the function wants an NSObject specifically, not some subclass, which is why you end up having to cast all the time.